src/webgpu/api/validation/capability_checks/limits/maxComputeWorkgroupStorageSize.spec.ts

```diff
@@ -0,0 +1,221 @@
+import { keysOf } from '../../../../../common/util/data_tables.js';
+import { assert } from '../../../../../common/util/util.js';
+import { roundDown } from '../../../../util/math.js';
+
+import { LimitValueTest, TestValue, kLimitBaseParams, makeLimitTestGroup } from './limit_utils.js';
+
+const limit = 'maxComputeWorkgroupStorageSize';
+export const { g, description } = makeLimitTestGroup(limit);
+
+const kSmallestWorkgroupVarSize = 4;
+
+const wgslF16Types = {
+  f16: { alignOf: 2, sizeOf: 2 },
+  'vec2<f16>': { alignOf: 4, sizeOf: 4 },
+  'vec3<f16>': { alignOf: 8, sizeOf: 6 },
+  'vec4<f16>': { alignOf: 8, sizeOf: 8 },
+  'mat2x2<f16>': { alignOf: 4, sizeOf: 8 },
+  'mat3x2<f16>': { alignOf: 4, sizeOf: 12 },
+  'mat4x2<f16>': { alignOf: 4, sizeOf: 16 },
+  'mat2x3<f16>': { alignOf: 8, sizeOf: 16 },
+  'mat3x3<f16>': { alignOf: 8, sizeOf: 24 },
+  'mat4x3<f16>': { alignOf: 8, sizeOf: 32 },
+  'mat2x4<f16>': { alignOf: 8, sizeOf: 16 },
+  'mat3x4<f16>': { alignOf: 8, sizeOf: 24 },
+  'mat4x4<f16>': { alignOf: 8, sizeOf: 32 },
+};
+
+const wgslBaseTypes = {
+  f32: { alignOf: 4, sizeOf: 4 },
+  i32: { alignOf: 4, sizeOf: 4 },
+  u32: { alignOf: 4, sizeOf: 4 },
+
+  'vec2<f32>': { alignOf: 8, sizeOf: 8 },
+  'vec2<i32>': { alignOf: 8, sizeOf: 8 },
+  'vec2<u32>': { alignOf: 8, sizeOf: 8 },
+
+  'vec3<f32>': { alignOf: 16, sizeOf: 12 },
+  'vec3<i32>': { alignOf: 16, sizeOf: 12 },
+  'vec3<u32>': { alignOf: 16, sizeOf: 12 },
+
+  'vec4<f32>': { alignOf: 16, sizeOf: 16 },
+  'vec4<i32>': { alignOf: 16, sizeOf: 16 },
+  'vec4<u32>': { alignOf: 16, sizeOf: 16 },
+
+  'mat2x2<f32>': { alignOf: 8, sizeOf: 16 },
+  'mat3x2<f32>': { alignOf: 8, sizeOf: 24 },
+  'mat4x2<f32>': { alignOf: 8, sizeOf: 32 },
+  'mat2x3<f32>': { alignOf: 16, sizeOf: 32 },
+  'mat3x3<f32>': { alignOf: 16, sizeOf: 48 },
+  'mat4x3<f32>': { alignOf: 16, sizeOf: 64 },
+  'mat2x4<f32>': { alignOf: 16, sizeOf: 32 },
+  'mat3x4<f32>': { alignOf: 16, sizeOf: 48 },
+  'mat4x4<f32>': { alignOf: 16, sizeOf: 64 },
+
+  S1: { alignOf: 16, sizeOf: 48 },
+  S2: { alignOf: 4, sizeOf: 16 * 7 },
+  S3: { alignOf: 12, sizeOf: 32 },
+};
+
+const wgslTypes = { ...wgslF16Types, ...wgslBaseTypes };
+type WGSLType = keyof typeof wgslTypes;
+const kWGSLTypes = keysOf(wgslTypes);
+
+function getModuleForWorkgroupStorageSize(device: GPUDevice, wgslType: WGSLType, size: number) {
+  assert(size % kSmallestWorkgroupVarSize === 0);
+  const { sizeOf, alignOf } = wgslTypes[wgslType];
+  const unitSize = Math.max(sizeOf, alignOf);
```

Comment:
Strictly I think you want `align(sizeOf, alignOf)` (from `util/math.ts`)

Reply:
done